### PR TITLE
Dance ensure characters referenced loaded

### PR DESCRIPTION
**Pull request by cpirich**
* This change enables static analysis of dance-party code. It requires a new version of `dance-party` that has not yet been published.
* Updated the dance loading animation logic to match this new method - previously, it was created with the `id` of `p5_loading`, which means that `p5` will remove it automatically as we exit the `setup()` phase. This only worked if we loaded our assets only once and within the `preload()` phase. The new version of `dance-party` does asset loading within `ensureSpritesAreLoaded()` after the `preload()` and `setup()` phases are complete. For compatibility with the existing `visualization` column layout, I've kept the loading animation `div` within the `divDance` div. But since it is within the `ProtectedVisualizationDiv`, we can't use the React `levelRunIsStarting` property to conditionally render it. Instead, we must always render it and toggle the `display` style between `flex` and `none` as we enter and exit `runIsStarting` mode (similar to how we handle disabling the `Run` button.

### CODE REVIEW

**Review by islemaster — APPROVED** [review]
:+1: I know we had some concerns about using regex for this being fragile re: block changes in the future, but this seems like a thing that would break in an obvious way for someone making block changes.

I'd like to rearrange things a little to remove the `charactersReferenced` instance variable, but otherwise this seems good.

File: apps/src/dance/Dance.js [hunk 1]
```diff
@@ -590,8 +592,22 @@ Dance.prototype.initInterpreter = function () {
...
+
+  // Process studentCode to determine which characters are referenced and update
+  // this.charactersReferenced array with the results:
+  this.charactersReferenced = [];
+  const charactersRegExp = new RegExp(/^.*makeNewDanceSprite(?:Group)?\([^"]*"([^"]*)[^\r\n]*/, 'gm');
+  let match;
+  while ((match = charactersRegExp.exec(studentCode))) {
+    const characterName = match[1];
+    if (!this.charactersReferenced.includes(characterName)) {
+      this.charactersReferenced.push(characterName);
```
Comment (islemaster) [inline apps/src/dance/Dance.js]: Let's use `Set` instead of doing our own duplicate-check.

Comment (cpirich) [reply apps/src/dance/Dance.js]: Yes, I considered that. Happy to see us change to that.

Comment (cpirich) [reply apps/src/dance/Dance.js]: Fixed to use `Set`. Also, per comment above, eliminated `charactersReferenced` as an instance variable. The array is now returned from `initInterpreter()`.

**Comment by islemaster** [comment]
Relevant write-up Chris sent me last night:

> I've changed how we load spritesheets and animations so that we can load only the assets needed for the characters referenced in the student's program.
> 
> Since p5 only initializes once (and therefore enters preload and setup once), we must now do asset loading without relying on the p5 preload phase (which through trickery tracks async asset loading functions and delays the exit of the preload phase and the start of the setup phase until after all of those async asset loading functions have completed.)
> 
> The fetch of `characters.json`, the spritesheet PNGs, and the construction of the `p5.play` animations now occur within an async function that must be completed before `play()` can be called. This function is called `ensureSpritesAreLoaded()`. It can be called without parameters (in which case all character assets are loaded) or with an array of character names that must be loaded.
> 
> The asset loading is additive. If a student's code is modified to reference additional characters, `ensureSpritesAreLoaded()` should be called again with a different list of characters. Any characters not yet loaded will be loaded before the Promise it returns is resolved.
> 
> We do not unload any animations at this time. We do not do any conditional loading of specific dance moves. We do cache the `characters.json` response in memory and do not request it more than once, much as we already cache each spritesheet PNG `Image` object and will not request those more than once.
> 
> Many new tests have been added to verify the behavior of `ensureSpritesAreLoaded()`. Existing tests have been updated to ensure that all sprites are loaded before calling `play()`. The test runner (`tape`) has been augmented with `tape-async` for unit tests to ensure that we can properly perform async waits within our tests.
> 
> The benefits are:
> 1. Code Studio is responsive and ready to run more quickly
> 2. The network payload is reduced for any program that doesn't use all characters (we only download what we need) so we can start dancing quickly when Run is pressed
> 3. The memory footprint is reduced for any program that doesn't use all characters.
> 4. Visiting a share page is lighter weight and faster, where we cannot expect to have the assets already in the browser cache. We now can start the dancers much more quickly (unless the shared program uses every character).

**Comment by cpirich** [comment]
Note: tests will not pass until `dance-party` is updated with the changes in https://github.com/code-dot-org/dance-party/pull/253 and published. Then, this PR will be updated to reference that new version of the `dance-party` package.

**Comment by islemaster** [comment]
We decided during triage today that we will not be accepting this change prior to Hour of Code.  We are still interested in taking it after Hour of Code.  We will revisit this on December 10th.

**Comment by islemaster** [comment]
Closing in favor of https://github.com/code-dot-org/code-dot-org/pull/26243 which is a strict superset of this change.

**Pull request closed without merging**